### ISSUE TITLE
re: neomake, an issue about loss of input control but works in vim

### ISSUE THREAD

**Issue by Earnestly** [neovim/neovim#7484]
Hi,

This is mostly for cross-reference to an issue I thought may be caused by neomake but testing found vim to function correctly.

Most of the information is here: https://github.com/neomake/neomake/issues/1704

The relevant parts are demonstrated in this short screen capture: https://ptpb.pw/hTnL.webm

As per the neomake issue, I have tried the latest nvim, latest libtermkit and latest libvterm.  Additionally I have also tried neovim with a clean configuration only adding vim-plug and neomake to replicate the problem.

This issue is present on termite (libvte) and xterm so I have ruled out the terminal emulator as the cause in this instance.

I'm really at a loss for what might cause this and I don't have any basis to attempt bisecting neovim.

Any pointers to what I can try, or any additional information you need, please ask.

Thanks

**Comment by justinmk** [comment]
@Earnestly Looks like control codes are being printed literally. Can you describe more precisely the other behavior symptoms?

**Comment by Earnestly** [comment]
As mentioned, this behaviour occurs in both xterm and termite (which is libvte based). I don't know of any GUI to test this in

The behaviour is difficult to describe as it's not purely printing control codes.  It almost seems like it reverts back to a vague form of readline behaviour, for example, `C-u` will erase the mess of characters but `C-a` and `C-e` are printed literally as `^A` and `^E`.

`Return` also correctly moves to the next line instead of printing `C-j` or `C-m`.  Additionally `Backspace` also works in the same manner as `C-u` except one character at a time.

I do suspect that [`spar`](https://github.com/kburtch/SparForte) is partially responsible for exposing this issue.  I may be able to build a static version of it so others can try since I don't expect people to build Ada very often.  Nevertheless it does exist in the [AUR](https://aur.archlinux.org/packages/sparforte-git/) if anyone is willing to build it locally.

For what it's worth, according to `ncursesw6-config --version` and I using version `6.0.20170902`.

**Comment by Earnestly** [comment]
It is strange that `spar` seems to be the only one causing this issue but I can't determine anything it might be doing wrong.  I have tried piping it to `| cat -A` and tried to launch it as its own command to see if it tries to provide a kind of REPL.  But neither happens and for all intents and purposes the output behaves identically to `gcc -fsyntax-only -x ada -c issue.sp`.

The only difference is if I change my `ft` to `ada` to run the above gcc command, it doesn't corrupt the input.

I'm really stumped by this...

**Comment by justinmk** [comment]
So somehow the neomake "maker" (spar?) is modifying the TTY or termios settings. Neomake doesn't do exactly the same steps for Vim, so one needs to find out what's different.

**Comment by Earnestly** [comment]
Good point about termios.  I'll dig around in the codebase and see if I can find anything.

As an additional check I ran it under `script -c 'spar -g -c issue.sp'` and the output is clean, e.g. `issue.sp:4:1: begin expected^M`

**Comment by Earnestly** [comment]
So this language seems to [call the `tput` tool](https://github.com/kburtch/SparForte/blob/master/src/spar_os-tty.adb#L82) to capture the escape sequences in a temporary file.

This is confirmed by `strace` but there doesn't seem to be any indication that a write occures with the escapes it collects.

I wrote a trivial parser for these and pulled it from the strace output to get this:

```
% strace spar -g -c issue.sp 2>&1 | ./parse-strace | paste -sd "" -
\33(B\33[m\33[1m\33[7m\33[J\33[K\33[A\33[C\7\33c\33[!p\33[?3;4l\33[4l\33>\33[H\33[2J42\n170\n#!/bin/spar\n\nprocedure issue is\nend issue;\n
```

While I have no evidence any of these are run, it does have some interesting bits in it.  E.g. for my current terminfo the`\33[2J` is `clear` and `\33[H` is `cub1`.  I don't think this is very helpful though as I tried to produce a dummy script which executed these commands to see if it could be used as replication but didn't get anywhere.

I've also tried wrapping `spar` whereby I store and restore the tty and that didn't help either.

:(

**Comment by justinmk** [comment]
Steps to reproduce mentioned at https://github.com/neomake/neomake/issues/1704#issuecomment-342309902

Maybe related: https://github.com/neovim/neovim/issues/6938

**Comment by Earnestly** [comment]
I'm not sure if https://github.com/neovim/neovim/issues/6938 is related because I can't see any `spar` process still running.
